File: anomalous-value-js/src/agent.js
const ethers = require('ethers');
const BigNumber = require('bignumber.js');
const RollingMath = require('rolling-math');
const {
  Finding, FindingSeverity, FindingType, getJsonRpcUrl,
} = require('forta-agent');

// Need to initialize a dictionary whose keys will be contract addresses the agent encounters
// and whose values are RollingMath objects tracking the ether values of transactions that
// touch the contract
const contractAddressesDict = {};

const providerObject = new ethers.providers.JsonRpcProvider(getJsonRpcUrl());

function provideHandleTransaction(contractAddresses, provider) {
  return async function handleTransaction(txEvent) {
    const findings = [];

    // skip if transaction is contract creation
    if (!txEvent.to) {
      return findings;
    }

    const value = new BigNumber(txEvent.transaction.value);

    // check if we've seen this address
    if (contractAddresses[txEvent.to]) {
      const average = contractAddresses[txEvent.to].getAverage();
      const standardDeviation = contractAddresses[txEvent.to].getStandardDeviation();

      // if the value is over 5 standard deviations from the mean and
      // we have a sample size of more than 40, report the finding
      if (value.isGreaterThan(average.plus(standardDeviation.times(5)))
          && contractAddresses[txEvent.to].getNumElements() > 40) {
        findings.push(
          Finding.fromObject({
            name: 'Abnormally High Value Transaction',
            description: `An abnormally high value of ${value} was sent to ${txEvent.to}`,
            alertId: 'AE-ANOMALOUS-VALUE',
            severity: FindingSeverity.Medium,
            type: FindingType.Suspicious,
            metadata: { contract: txEvent.to },
          }),
        );
      }
    } else if (await provider.getCode(txEvent.to) !== '0x') {
      // address is a contract we haven't seen before, initialize it
      // eslint-disable-next-line no-param-reassign
      contractAddresses[txEvent.to] = new RollingMath(1000);
    } else {
      // externally owned account, return
      return findings;
    }

    // add transaction to contractAddresses dictionary
    contractAddresses[txEvent.to].addElement(value);
    return findings;
  };
}

module.exports = {
  provideHandleTransaction,
  handleTransaction: provideHandleTransaction(contractAddressesDict, providerObject),
};

File: malicious-addr-py/src/agent.py
from forta_agent import Finding, FindingType, FindingSeverity

from src import malicious_addrs


def handle_transaction(transaction_event):
    """
    Check to see if the malicious address was involved with a transaction.
    Return an empty list if the malicious address is not involved and does not
    trigger an alert
    """
    addresses = transaction_event.addresses

    # If no malicious addresses are involved in the transaction, no alert should be raised
    for addr in addresses:
        if addr in malicious_addrs.addrs:
            break
    else:
        return []

    alerts = []
    bad_addrs = []
    # If the malicious address is involved with the transaction send an alert
    for addr in addresses:
        if addr in malicious_addrs.addrs:
            bad_addrs.append(addr)
    else:
        if bad_addrs:
            alert = Finding(
                {
                    "name": "Malicious Address Detected",
                    "description": "Malicious address is involved with a transaction",
                    "alert_id": "AE-MALICIOUS-ADDR",
                    "type": FindingType.Suspicious,
                    "severity": FindingSeverity.Medium,
                    "metadata": {
                        "from": transaction_event.transaction.from_,
                        "to": transaction_event.transaction.to,
                        "amount": transaction_event.transaction.value,
                        "malicious_addresses": bad_addrs,
                    },
                }
            )
            alerts.append(alert)

    return alerts
